src/CommunityToolkit.Mvvm.SourceGenerators/Helpers/ImmutableArrayBuilder{T}.cs

```diff
@@ -173,6 +190,48 @@ public void Dispose()
             }
         }
 
+        /// <inheritdoc/>
+        void ICollection<T>.Clear()
+        {
+            throw new NotSupportedException();
+        }
+
+        /// <inheritdoc/>
+        bool ICollection<T>.Contains(T item)
+        {
+            throw new NotSupportedException();
+        }
+
+        /// <inheritdoc/>
+        void ICollection<T>.CopyTo(T[] array, int arrayIndex)
+        {
+            Array.Copy(this.array!, 0, array, arrayIndex, this.index);
+        }
+
+        /// <inheritdoc/>
+        IEnumerator<T> IEnumerable<T>.GetEnumerator()
+        {
+            T?[] array = this.array!;
+            int length = this.index;
+
+            for (int i = 0; i < length; i++)
+            {
+                yield return array[i]!;
```

Comment:
Wouldn't this allocate a state machine? Why is that better than allocating the array?

Reply:
Well, the array would just be on top of that. As in, if we create the array and pass that to Roslyn, it will then enumerate it anyway (as an `IEnumerable<T>`), so we'll still allocate an enumerator for that here. So this way instead we can at least skip the array entirely (which is just intermediate and wasted) and just enumerate directly 🙂
This also matches what you can do with the normal `ImmutableArray<T>.Builder` type.

Reply:
Makes sense!